plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java

```diff
@@ -184,17 +194,35 @@ public ConnectorPageSource createPageSource(
                 .map(IcebergColumnHandle.class::cast)
                 .collect(toImmutableList());
 
-        Map<Integer, Optional<String>> partitionKeys = split.getPartitionKeys();
+        HdfsContext hdfsContext = new HdfsContext(session);
+        FileIO fileIO = fileIoProvider.createFileIo(hdfsContext, session.getQueryId());
+        TrinoDeleteFilter deleteFilter = new TrinoDeleteFilter(
+                new DummyFileScanTask(split.getPath(), split.getDeletes()),
+                SchemaParser.fromJson(table.getSchemaJson()),
+                icebergColumns,
+                fileIO);
 
-        List<IcebergColumnHandle> regularColumns = columns.stream()
-                .map(IcebergColumnHandle.class::cast)
+        Map<Integer, Optional<String>> partitionKeys = split.getPartitionKeys();
+        ImmutableList.Builder<IcebergColumnHandle> columnsForDataPageSource = ImmutableList.builder();
+        icebergColumns.stream()
+                .filter(column -> !partitionKeys.containsKey(column.getId()))
+                .forEach(columnsForDataPageSource::add);
+        List<IcebergColumnHandle> deleteFilterRequirements = getColumns(deleteFilter.requiredSchema(), typeManager);
+        deleteFilterRequirements.stream()
+                .filter(deleteRequiredColumn -> !icebergColumns.contains(deleteRequiredColumn))
```

Comment:
Since we are doing a distinct here:
```
 List<IcebergColumnHandle> allColumns = Streams.concat(icebergColumns.stream(), deleteFilterRequirements.stream())
                .distinct()
                .collect(toImmutableList());
```, 
we can remove this filter on the stream.